### ISSUE TITLE
Allow Parameters to be Passed in to InModuleScope blocks

### ISSUE THREAD

**Issue by vexx32** [pester/Pester#1158]
/cc @nohwnd 

## 1. General summary of the issue

If a test file depends on a helper file for some static data that is used during testing, it can be very difficult to use with InModuleScope because often such files are located at some fixed location _relative to the test script_; frequently alongside it or in a subfolder.

## 3. Expected Behavior

InModuleScope should expose some parameters to allow passing items in, for example the path of the test script.

## 4.Current Behavior

User is typically forced to assess where tests are _likely_ to be relative to ModureBase or some such, which is _usually_ accurate but no necessarily so.

## 5. Possible Solution

Add a parameter to InModuleScope that accepts optional parameters which are provided to the script block, similar to what happens with this native syntax:

```powershell
& (Get-Module MyModule) {
    param($param1, $param2) 

    Write-Verbose $Param1 $Param2
} $PSScriptRoot (Get-Date)
```

That last line might require a comma, I don't recall.

## 6. Context

In working with testing some of the functions in PSKoans I have need to ensure static test scripts are appropriately tested with internal module helper functions and do not throw unusual (non-pester) errors that might be indicative of a problem in the module functionality.

**Comment by nohwnd** [comment]
For the paths you can use `$PSScriptRoot` or `$PSCommandPath` to get the test file path, and `$script:PSScriptRoot` or `$script:PSCommandPath` to get the module path. But I mention this mostly for completeness if someone reads this later. 

Do you have some more concrete practical examples, so it's easier to think about?

**Comment by vexx32** [comment]
Not in `InModuleScope` you can't, at least for my experimentation. It returns the same path as `(Get-Module ModuleName).ModuleBase` for the module you're in the scope of. Not sure about `PSCommandPath` but I suspect that will turn out to be similar.

**Comment by nohwnd** [comment]
Then you might be doing something wrong, I tried this in multiple places (and just now on pwsh 6 on MacOS) and it outputs path to where my test script is, and path to where PSReadLine is.

```powershell 
Describe "a" { 
    It "b" { 
        InModuleScope PSReadLine { 
            Write-Host $script:PSScriptRoot 
            Write-Host $PSScriptRoot
        } 
    } 
}
```

**Comment by FGS-BMarkham** [comment]
I think I'm running into the same limitation. I think the scenario could be described as mixing white-box and black-box testing at the same time.

For example, I'm writing tests around a module named `PNPStuff` and I want to test its exports and then move inward to test its insides. To facilitate such, I've written two helper functions:
```
    Function AssertFunctionExists([Parameter(Mandatory = $true)] [string] $functionName, [ScriptBlock] $And = $null) {
        Describe "Function $functionName" {
            $command = Get-Command -Module PNPStuff $functionName -ErrorAction SilentlyContinue

            It 'Exists As Command' {
                $command | Should -Not -BeNullOrEmpty
            }

            It 'Is A Function' {
                $command.CommandType | Should -Be Function
            }

            if ($And) {
                Invoke-Expression $And
            }
        }
    }

    Function AssertFunctionExistsInModuleScope([Parameter(Mandatory = $true)] [string] $functionName, [ScriptBlock] $And = $null) {
        InModuleScope PNPStuff {
            Describe "Function $functionName" {
                $command = Get-Command -Module PNPStuff $functionName -ErrorAction SilentlyContinue
    
                It 'Exists As Command' {
                    $command | Should -Not -BeNullOrEmpty
                }
    
                It 'Is A Function' {
                    $command.CommandType | Should -Be Function
                }
    
                if ($And) {
                    Invoke-Expression $And
                }
            }
        }
    }
```

Aside from me not being able to DRY them (I'm running into a problem that is a different conversation for a different day), I would expect these to behave the same. Instead, the latter function (which uses `InModuleScope`) isn't able to pass data parameters into the scriptblock - it acts as if I called `Describe "Function "`.

**Comment by nohwnd** [comment]
nohwnd [3:17 PM]
I am now facing a dillema. When adding a mock with -ModuleName parameter there are multiple places where a mock can run. It can run inside of the module, but then it loses access to the variables defined in the test. It can run in the test scope, but then it loses access to variables in the module scope. How this is done in v4, the mock uses a mix of both, the parameter filter runs in the test, but the MockWith runs in the module scope. I'd like to unify this, but I don't want to choose incorrectly, and I don't want to make it too difficult to use. One of the proposals is to allow for explicit parameter passing to mocks, but then using parameter filters would be quite unwieldy, eg. `-ParameterFilter { $Name -eq $NameFromTest } -Context @{ NameFromTest = $Name }`, in parameter filters we have the values, and I assume it is rare to filter on the data coming  from the module script scope. On the other hand in the body, stuff like module path, or other module functions might be needed, this is where I don't know. My mocking  is usually very simple and used for unit tests :)) how do you use your mocks?


chrisdent [3:19 PM]
+1 for "My mocking  is usually very simple and used for unit tests"

nohwnd [3:20 PM]
This issue calls for making everything run in the test scope, and just putting the hook in the module scope. https://github.com/pester/Pester/issues/204 and here a proposal from Joel with explicit parameters: https://github.com/pester/Pester/issues/1158
GitHub
Mock -ModuleName scoping issues · Issue #204 · pester/Pester
Mock -ModuleName is causing the MockWith script block to be executed inside of the module, instead of out in the test code&#39;s scope which is where every other parameter filter and mock script bl...
GitHub
Allow Parameters to be Passed in to InModuleScope blocks · Issue #1158 · pester/Pester
/cc @nohwnd 1. General summary of the issue If a test file depends on a helper file for some static data that is used during testing, it can be very difficult to use with InModuleScope because ofte...
Re-reading Joels issue it also seems that the problem is that the test scope is unavailable, not that they are missing the module scope data. So maybe running everything in the test scope is the correct way all along?

chrisdent [3:25 PM]
I tend to use InModuleScope a great deal for any tests requiring access to module-internal functions or variables

nohwnd [3:28 PM]
Yeah so that would make the whole test run in the module scope, making both the module scope and the test scope variables accessible.

chrisdent [3:29 PM]
yep
I suppose in theory it would be better to act in test scope. It's just that feels like a lot more effort :slightly_smiling_face:

nohwnd [3:30 PM]
Or you could do `It 'abc' { $moduleState = InModuleScope -ModuleName "Module1" { $script:internalPath } ... rest of the test }`
for me the usage is: I have a module. I am testing a public function that depends on an internal function. I want to mock the internal function, but still access the public function from the public side.

chrisdent [3:33 PM]
same for me really, probably have to review some of my tests.

nohwnd [3:33 PM]
Then I will aim for the 204 way, where everything runs in the test scope. That should also fix Joels problem without adding more parameters.
@chrisdent please do, in case you find something that would not work with this approach :))

Putting a copy of a slack discussion here so I can refer back to it. The conclusion is to follow what is outlined here. It should not be that hard now that aliases are used for mocks, we just need to lookup the function info from the module because it might be just internal, but then hook it to a bootstrap function that will execute in the test scope.

**Comment by fourpastmidnight** [comment]
Well, I've since gotten around this issue in several ways. First, I was finding that `InModuleScope` didn't work well at all for Gherkin style tests, especially when some steps are reused and sometimes need to be called in the module scope, while other times it does not. You cant' have both. So, I've gotten around this:

I have the following structure for my test suite:

```text
ProjectName
  ├─ ProjectName
  │    ├─ Private
  │    ├─ Public
  │    ├─ ProjectName.psd1
  │    └─ ProjectName.psm1
  └─ specs
        └─ features
            ├─ step_definitions
            └─ support
                ├─ Environment.ps1
                ├─ GlobalVars.ps1
                ├─ hooks.Steps.ps1
                └─ Transforms.ps1
```

So, Pester Gherkin doesn't support Transforms (yet 😉 ). Global variables I use sparingly, but are sometimes required due to scoping issues in Pester Gherkin. But the real meat here is in `hooks.Steps.ps1` and `Environment.ps1`. (It should be noted that in the v5 updates to Gherkin I'm proposing, the `Environment.ps1` file will be a "first-class citizen" of the Gherkin test suite, working much like Ruby Cucumber uses such a file (which is named `env.rb`).

### hooks.Steps.ps1

```powershell
Set-StrictMode -Version Latest

$World = Get-Item (Join-Path $PSScriptRoot Environment.ps1)
if ($World) {
    . $World.FullName
}

AfterEachFeature {
    ${Env:BHProjectName}, 'MyProjectName.Private' |
        Get-Module |
        Remove-Module -Force
}
```

### Environment.ps1

```powershell
Set-StrictMode -Version Latest

# ! Dot source any other files here in the support folder which initialize the test environment
# ! (known as "the World" in Ruby Cucumber. By default, all *.ps1 files in the support directory
# ! imported.
(Get-ChildItem $PSScriptRoot -File -Recurse -Filter *.ps1) |
    Where-Object { $_.Name -notin 'hooks.Steps.ps1','Environment.ps1' } |
    Select-Object -ExpandProperty FullName |
    ForEach-Object { . $_ }

# Gherkin-style tests really can't make use of Pester's InModuleScope due to the nature of Gherkin-style
# tests' test structure. Steps which sometimes need to be in module scope and other times do not are
# problematic for Pester. And in Cucumber, steps are supposed to be global. So, to get around this,
# create a dynamic module containing all the public and private functions so that we can test them all.
if (!(Get-Module "${Env:BHProjectName}.Test")) {
    New-Module -Name "${Env:BHProjectName}.Test" {
        ${Env:BHModulePath} |
            Get-ChildItem -Directory |
            Get-ChildItem -File -Recurse -Filter *.ps1 |
            ForEach-Object {
                Write-Verbose "Dot sourcing '$($_.Name)'"
                . ($_.FullName)
            }
    } | Import-Module -Global -Force

    Get-ChildItem "${Env:BHModulePath}" -Filter *.types.ps1xml -Recurse |
        Select-Object -ExpandProperty FullName |
        Update-TypeData

    Get-ChildItem "${Env:BHModulePath}" -Filter *.format.ps1xml -Recurse |
        Select-Object -ExpandProperty FullName |
        Update-FormatData
}

# *NOTE: Because Pester does not appear to scope hooks and Step Definition script blocks as one
# *      might expect, any functions you want to be art of the environment during step definition
# *      execution must be placed in the global scope.

function Global:New-MockFor {
    Param(
        [Parameter(Position = 0, Mandatory = $True)]
        [string]$CommandName,

        [Parameter(Position = 1)]
        [ScriptBlock]$Return = {},

        [Parameter(Position = 2)]
        [Alias('WhenParametersMatch','WhenParametersSatisfy','WhereParametersSatisfy','WhenGivenParameters','WhenGivenParametersThatSatisfy')]
        [ScriptBlock]$WhereParametersMatch = {$True}

        [string]$InModule = "${Env:BHProjectName}.Test"
    )

    Mock $CommandName -ModuleName $InModule -ParameterFilter $WhereParametersMatch -MockWith $Return
}

function Global:Test-MockCalled {
    Param (
        [Parameter(Position = 0, Mandatory = $True)]
        [String]$CommandName,

        [Parameter(Position = 2)]
        [Alias('WhenParametersMatch','WhenParametersSatisfy','WhereParametersSatisfy','WhenGivenParameters','WhenGivenParametersThatSatisfy')]

        [string]$InModule = "${ENV:BHProjectName}.Test",

        [int]$Times = 1

        [switch]$Exactly
    )

    Assert-MockCalled $CommandName -InModule $InModule -ParameterFilter $WhereParametersMatch -Times $Times -Exactly:$Exactly
}
```

The `${Env:BH*}` variables come from the BuildHelper module by @ramblingcookiemonster. I've built a whole "development environment" in VS code that takes care of building my module, running the tests, creating documentation etc. It's really pretty sweet. Anyway, with the above two files, I've pretty much eliminated my need for `InModuleScope`.

To get around passing parameters into mocks, I create closures for the `$Return` parameter of my `New-MockFor` wrapper around Pester's `Mock`. Here's an example from a module I recently wrote for work (this code is safe to post here, it doesn't really show much of anything "proprietary" or "secret"):

```powershell
New-MockFor Get-deploymentTargetResource -Return (
    & {
        Param([PSCustomObject]$DeploymentTargetResource)

        $Target = $DeploymentTargetResource

        {
            $ShouldReturnATarget = $False
            foreach ($role in 'some-role-i-care-about','another-role') {
                $ShouldReturnATarget = $ShouldReturnATarget -or $Script:Target.Roles -contains $role
            }

            if ($ShouldReturnATarget) {
                @($Script:Target)
            } else {
                @()
            }
        }.GetNewClosure()
    } $Script:Context.DeploymentTarget
) -WhenGivenParametersThatSatisfy {
        (
            'a-role-that-i-care-about','another-role' | ForEach-Object -Begin {
                $Result = $True
            } -Process {
                $Result = $Result -and $Roles -contains $_
            } -End {
                $Result
            }
        ) -and $EnvironmentId -eq 'Environments-1'
}
```

The details aren't important (I was creating a module to interoperate with Octopus Deploy--I didn't like any of the ones that are out there). But you can (in my opinion) see how nicely the `New-MockFor` invocation reads:

`New-MockFor myCommand -returns { ... } -WhenGivenParametersThatSatisfy { ... } [-InModule 'foo' <# if needed, but it defaults to the private module created in Environment.ps1 #>]`

The only thing that's wonky right now is the parameter filter aspect of Mocks. Because they aren't executed in the step's scope (or even in the test runner's scope at all!) you have _NO_ access to _ANY_ data in the test. So when I need data from the test inside the parameter filter, I have to resort to sting comparisons only so that I can "stringify" the paremeter filter script block and use `[ScriptBlock]::Create('...')` to pass it to the mock.

**Comment by nohwnd** [comment]
@fourpastmidnight  thanks, I just don't think I can use any of that :)

**Comment by fourpastmidnight** [comment]
@nohwnd Agreed. I don't think this would become part of Pester proper in any way. But I think for Gherkin-style tests, it will always be challenging to have things scoped "correctly". So, I posted this mainly so that if anyone else is looking for ways to "have their cake and eat it, too", this is how I'm able to do that.

IOW, maybe we don't need to worry about the "scoping issues" so much for Gherkin. For me, the biggest pain point is the scope of Parameter Filters. They sort of need to be in both the test script block scope _and_ the mocked command's scope, and I just don't see how that could be done.

Is it possible to "inject" values into the state created for the parameter filter?

**Comment by nohwnd** [comment]
@fourpastmidnight yes. I already do that in v5, and here is an isolated example of how to do that: https://gist.github.com/nohwnd/3ca1c5f2240ee187303867ddb7422665

**Comment by fourpastmidnight** [comment]
I'm still a bit weak on scoping and sessions in PowerShell (but getting better!)--so I still don't grok that example _completely_, but I'm realy glad to see this has been worked on. I'm looking forward to Pester v5! 😄

**Comment by nohwnd** [comment]
Fixed in #1281 bodies and parameter filters now run in the test scope.
